magma/value_utils.py

```diff
@@ -0,0 +1,107 @@
+import dataclasses
+import typing
+
+from magma.ref import Ref, TupleRef, ArrayRef
+from magma.type_utils import isprotocol, isdigital, isbits, isarray, istuple
+
+
+def _unwrap_protocol_value(value):
+    T = type(value)
+    if isprotocol(T):
+        return value._get_magma_value_()
+    return value
+
+
+@dataclasses.dataclass(frozen=True)
+class _ValueWrapper:
+    raw: typing.Any
+    children: typing.List[typing.Any]
+    name: str
+
+
+def _wrap_value(value):
+    if isinstance(value, _ValueWrapper):
+        return value
+    value = _unwrap_protocol_value(value)
+    T = type(value)
+
+    if isdigital(T):
+        return _ValueWrapper(value, [], "Digital")
+    if isbits(T):
+        return _ValueWrapper(value, [], "Bits")
+    if isarray(T):
+        return _ValueWrapper(value, [t for t in value], "Array")
+    if istuple(T):
+        return _ValueWrapper(value, [t for t in value], "Tuple")
```

Comment:
Perhaps we could add a utility like `isdigitalval` or `isinstdigital` or something rather than having to fetch `type(value)` every time? Honestly this isn't so bad though and packing 3 words without underscores kinda hurts my eyes (`is_digital_value` wouldn't be so bad, but then we'd probably want to change everything to be consistent and it's not consistent with isinstance/issubclass)

Reply:
an alternative would be for these `is<x>` methods to be general for both values and types, so like:

```python
def _isdigital(T):
    return issubclass(T, Digital)

def isdigital(value_or_T):
    try:
        return _isdigital(value_or_T)
    except TypeError:
        pass
    value_or_T = type(value_or_T)
    try:
        return _isdigital(value_or_T)
    except TypeError:
        pass
    return False
```

then

```python
assert isdigital(Bit)
assert isdigital(Bit())
```

It is a bit overloaded and not clear but logically it just applies the same logic to the value or the type of the value. Let me know how it reads.

Reply:
Leaving as is for now.